Enable LLDB to load large dSYM files.

clayborg opened the pull request:
llvm-dsymutil can produce mach-o files where some sections in __DWARF exceed the 4GB barrier and subsequent sections in the dSYM will be inaccessible because the mach-o section_64 structure only has a 32 bit file offset. This patch enables LLDB to load a large dSYM file by figuring out when this happens and properly adjusting the file offset of the LLDB sections.

I was unable to add a test as obj2yaml and yaml2obj are broken for mach-o files and they can't convert a yaml file back into a valid mach-o object file. Any suggestions for adding a test would be appreciated.

clayborg committed:
Enable LLDB to load large dSYM files.

llvm-dsymutil can produce mach-o files where some sections in __DWARF exceed the 4GB barrier and subsequent sections in the dSYM will be inaccessible because the mach-o section_64 structure only has a 32 bit file offset. This patch enables LLDB to load a large dSYM file by figuring out when this happens and properly adjusting the file offset of the LLDB sections.

I was unable to add a test as obj2yaml and yaml2obj are broken for mach-o files and they can't convert a yaml file back into a valid mach-o object file. Any suggestions for adding a test would be appreciated.

clayborg commented:
llvm-dsymutil will assert when asserts are enabled when creating large dSYM files that exhibit this bevavior. If asserts are disabled, then llvm-dsymutil can produce such binaries.

DataCorrupted commented on lldb/source/Plugins/ObjectFile/Mach-O/ObjectFileMachO.cpp:
```diff
@@ -1697,6 +1701,14 @@ void ObjectFileMachO::ProcessSegmentCommand(
+    // If this section overflows a 4GB barrier, then we need to adjust any
+    // subsequent the section offsets.
+    if (is_dsym && ((uint64_t)sect64.offset + sect64.size) >= UINT32_MAX)
+      section_offset_adjust += 0x100000000ull;
```
Imagine there are two sections:

```
__sec_a, size 8GB, offset 0
__sec_b, size 100, offset 0
```

Current `0x100000000ull` (4G) increment will think that `__sec_b` starts at 4G instead of 8G.

jasonmolenda commented:
FWIW a month ago I had a PR https://github.com/llvm/llvm-project/pull/159849 where lldb loads some mach-o structures into llvm::MachO representations, and then adds a virtual memory slide to them which could exceed 4GB.  I made local copies of the structures (e.g. `DyldInfoCommandLargeOffsets`) with all the same field names, and a ctor that takes the llvm::MachO object and populates all the fields.  Then all uses of the variable of my DyldInfoCommandLargeOffsets type behave like the llvm::MachO structure.  It would remove the side variable tracking the section offset.

jasonmolenda commented on lldb/source/Plugins/ObjectFile/Mach-O/ObjectFileMachO.cpp:
```diff
@@ -1697,6 +1701,14 @@ void ObjectFileMachO::ProcessSegmentCommand(
+    // If this section overflows a 4GB barrier, then we need to adjust any
+    // subsequent the section offsets.
+    if (is_dsym && ((uint64_t)sect64.offset + sect64.size) >= UINT32_MAX)
+      section_offset_adjust += 0x100000000ull;
```
```
    const uint64_t section_file_offset = sect64.offset + section_offset_adjust;
     if (is_dsym && ((uint64_t)sect64.offset + sect64.size) >= UINT32_MAX)
      section_offset_adjust += 0x100000000ull;
```

I wouldn't check `is_dsym`.  If this is correct, it is correct for any mach-o file that exceeds 4GB; the fact that we only see this in practice in dSYMs doesn't mean we should restrict it here.

clayborg committed:
Address review comments.

- Fix a case where a section can be larger that 4GB
- Fix comments to be a bit more clear
- Don't only do this for dSYM files

clayborg commented:
All inline suggestions have been fixed. I also tried to make a minimal yaml file that could create a large enough file, but it complains when trying to emit the file as it runs into the 32 bit section offset limiation and refuses to make a bad mach-o file. So no easy way to test this with yaml.

clayborg committed:
Run clang format.

ellishg commented on lldb/source/Plugins/ObjectFile/Mach-O/ObjectFileMachO.cpp:
```diff
@@ -1697,6 +1701,15 @@ void ObjectFileMachO::ProcessSegmentCommand(
+    // for the file offset of the section contents, so we need to track and
+    // sections that overflow and adjust the offsets accordingly.
+    const uint64_t section_file_offset = sect64.offset + section_offset_adjust;
+    const uint64_t end_section_offset = (uint64_t)sect64.offset + sect64.size;
```
Do we need to convert `sect64.offset` to `uint64_t`? And if we do, why not do the same on the line before?
```suggestion
    const uint64_t section_file_offset = sect64.offset + section_offset_adjust;
    const uint64_t end_section_offset = sect64.offset + sect64.size;
```

drodriguez reviewed (commented):
> I also tried to make a minimal yaml file that could create a large enough file, but it complains when trying to emit the file as it runs into the 32 bit section offset limiation and refuses to make a bad mach-o file. So no easy way to test this with yaml.

Since it is difficult to generate an invalid Mach-O, I wonder if in this case a compressed file that's mostly zeros can be added to the repository. Uncompress during the test, hope the contributor has the 4 GBs to spare, then delete as one of the last things on the test (because nobody wants a silly 4 GB in their system).

clayborg committed:
Cast to uint64_t to be clear what we are doing.

clayborg committed:
clang format.

clayborg committed:
Add a test.

The binary is as minimal as possible and it contains 1 segment named "__DWARF" with 3 sections:

FILE OFF    INDEX ADDRESS            SIZE               OFFSET     ALIGN      RELOFF     NRELOC     FLAGS      RESERVED1  RESERVED2  RESERVED3  NAME
=========== ===== ------------------ ------------------ ---------- ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------------------
0x00000068: [  1] 0x00000000fffffff0 0x0000000000000020 0xfffffff0 0x00000002 0x00000000 0x00000000 0x00000001 0x00000000 0x00000000 0x00000000 __DWARF.__debug_abbrev
0x000000b8: [  2] 0x0000000100000010 0x0000000200000000 0x00000010 0x00000002 0x00000000 0x00000000 0x00000001 0x00000000 0x00000000 0x00000000 __DWARF.__debug_info
0x00000108: [  3] 0x0000000300000010 0x0000000000000020 0x00000010 0x00000002 0x00000000 0x00000000 0x00000001 0x00000000 0x00000000 0x00000000 __DWARF.__debug_line

The file offsets should be parsed correctly by LLDB as:

__debug_abbrev file_offset=0x00000000fffffff0
__debug_info file_offset=0x0000000100000010
__debug_line file_offset=0x0000000300000010
